File: arp_parser.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <stdlib.h>
#include <windows.h>

struct arp_packet {
    uint16_t htype;
    uint16_t ptype;
    uint8_t hsize;
    uint8_t psize;
    uint16_t op;
    uint8_t sha[6];
    uint8_t spa[4];
    uint8_t tha[6];
    uint8_t tpa[4];
};

uint8_t test_arp_packet[] = {
    0x00, 0x01, 0x08, 0x00, 0x06, 0x04, 0x00, 0x01,
    0x08, 0x00, 0x27, 0x12, 0x34, 0x56, 0xC0, 0xA8,
    0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0xC0, 0xA8, 0x01, 0x02
};

uint16_t ntoh_16(const uint8_t *data) {
    return (data[0] << 8) | data[1];
}

int parse_arp(const uint8_t *data, size_t len, struct arp_packet *packet) {
    if (data == NULL || packet == NULL) {
        return -1;
    }
    
    if (len < 28) {
        return -2;
    }
    
    packet->htype = ntoh_16(data);
    packet->ptype = ntoh_16(data + 2);
    packet->hsize = data[4];
    packet->psize = data[5];
    packet->op = ntoh_16(data + 6);
    
    memcpy(packet->sha, data + 8, 6);
    memcpy(packet->spa, data + 14, 4);
    memcpy(packet->tha, data + 18, 6);
    memcpy(packet->tpa, data + 24, 4);
    
    if (packet->hsize != 6 || packet->psize != 4) {
        return -3;
    }
    
    return 0;
}

void mac_to_string(const uint8_t *mac, char *buffer) {
    sprintf(buffer, "%02x:%02x:%02x:%02x:%02x:%02x",
            mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
}

void ip_to_string(const uint8_t *ip, char *buffer) {
    sprintf(buffer, "%d.%d.%d.%d", ip[0], ip[1], ip[2], ip[3]);
}

const char* opcode_to_string(uint16_t op) {
    switch(op) {
        case 1: return "REQUEST (1)";
        case 2: return "REPLY (2)";
        default: return "UNKNOWN";
    }
}

void print_arp(const struct arp_packet *packet) {
    if (packet == NULL) {
        printf("Ошибка: NULL указатель на пакет\n");
        return;
    }
    
    char mac_buffer[18];
    char ip_buffer[16];
    
    printf("\n=== ARP Packet ===\n");
    printf("Hardware type: %u (0x%04x)\n", packet->htype, packet->htype);
    printf("Protocol type: 0x%04x", packet->ptype);
    
    if (packet->ptype == 0x0800) {
        printf(" (IPv4)");
    }
    printf("\n");
    
    printf("Hardware size: %u байт\n", packet->hsize);
    printf("Protocol size: %u байт\n", packet->psize);
    printf("Opcode: %s\n", opcode_to_string(packet->op));
    
    mac_to_string(packet->sha, mac_buffer);
    printf("Sender MAC: %s\n", mac_buffer);
    
    ip_to_string(packet->spa, ip_buffer);
    printf("Sender IP: %s\n", ip_buffer);
    
    mac_to_string(packet->tha, mac_buffer);
    printf("Target MAC: %s\n", mac_buffer);
    
    ip_to_string(packet->tpa, ip_buffer);
    printf("Target IP: %s\n", ip_buffer);
    
    if (packet->hsize == 6 && packet->psize == 4) {
        printf("[OK] Формат Ethernet/IPv4\n");
    }
    printf("==================\n");
}

void run_tests() {
    printf("\n=== ЗАПУСК ТЕСТОВ ===\n");
    
    // Тест 1: Нормальный пакет
    printf("\nТест 1: Нормальный пакет\n");
    struct arp_packet packet1;
    int result = parse_arp(test_arp_packet, sizeof(test_arp_packet), &packet1);
    if (result == 0) {
        printf("[OK] Успех\n");
        print_arp(&packet1);
    } else {
        printf("[ERROR] Ошибка: код %d\n", result);
    }
    
    // Тест 2: NULL указатель
    printf("\nТест 2: NULL указатель на данные\n");
    result = parse_arp(NULL, 28, &packet1);
    printf("Ожидаемый код: -1, Полученный код: %d - %s\n", 
           result, result == -1 ? "[OK]" : "[ERROR]");
    
    // Тест 3: Недостаточная длина
    printf("\nТест 3: Недостаточная длина (20 байт)\n");
    result = parse_arp(test_arp_packet, 20, &packet1);
    printf("Ожидаемый код: -2, Полученный код: %d - %s\n", 
           result, result == -2 ? "[OK]" : "[ERROR]");
    
    // Тест 4: Создание своего ARP-ответа
    printf("\nТест 4: Создание своего ARP-ответа\n");
    uint8_t custom_packet[] = {
        0x00, 0x01, 0x08, 0x00, 0x06, 0x04, 0x00, 0x02,
        0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff,
        0xc0, 0xa8, 0x01, 0x64,
        0x11, 0x22, 0x33, 0x44, 0x55, 0x66,
        0xc0, 0xa8, 0x01, 0x01
    };
    
    struct arp_packet packet2;
    result = parse_arp(custom_packet, sizeof(custom_packet), &packet2);
    if (result == 0) {
        printf("[OK] Успех\n");
        print_arp(&packet2);
    } else {
        printf("[ERROR] Ошибка: код %d\n", result);
    }
}

int main() {
    // Устанавливаем кодировку UTF-8 для консоли Windows
    SetConsoleOutputCP(CP_UTF8);
    
    printf("ARP Parser Demo\n");
    printf("===============\n");
    
    // Разбор тестового пакета
    printf("\n--- Разбор тестового пакета из задания ---\n");
    struct arp_packet packet;
    int result = parse_arp(test_arp_packet, sizeof(test_arp_packet), &packet);
    
    if (result == 0) {
        print_arp(&packet);
        
        // Проверка соответствия примеру
        printf("\nПроверка по заданию:\n");
        char ip_buffer[16];
        ip_to_string(packet.spa, ip_buffer);
        printf("Sender IP: %s (должен быть 192.168.1.1) - %s\n", 
               ip_buffer, strcmp(ip_buffer, "192.168.1.1") == 0 ? "[OK]" : "[ERROR]");
        
        ip_to_string(packet.tpa, ip_buffer);
        printf("Target IP: %s (должен быть 192.168.1.2) - %s\n", 
               ip_buffer, strcmp(ip_buffer, "192.168.1.2") == 0 ? "[OK]" : "[ERROR]");
    } else {
        printf("Ошибка парсинга: код %d\n", result);
    }
    
    // Запуск дополнительных тестов
    run_tests();
    
    printf("\nНажмите Enter для выхода...");
    getchar();
    
    return 0;
}

File: arp_parser.erl
-module(arp_parser).
-export([parse/1, print_arp/1, test/0]).

-record(arp_packet, {
    htype,      % hardware type (2 байта)
    ptype,      % protocol type (2 байта)
    hsize,      % hardware size (1 байт)
    psize,      % protocol size (1 байт)
    op,         % opcode (2 байта)
    sha,        % sender MAC (6 байт)
    spa,        % sender IP (4 байта)
    tha,        % target MAC (6 байт)
    tpa         % target IP (4 байта)
}).

% Тестовый пакет из задания
-define(TEST_PACKET, <<
    16#00, 16#01, 16#08, 16#00, 16#06, 16#04, 16#00, 16#01,
    16#08, 16#00, 16#27, 16#12, 16#34, 16#56, 16#C0, 16#A8,
    16#01, 16#01, 16#00, 16#00, 16#00, 16#00, 16#00, 16#00,
    16#C0, 16#A8, 16#01, 16#02
>>).

% Основная функция парсинга
parse(Binary) when is_binary(Binary) ->
    case byte_size(Binary) of
        28 ->
            parse_arp_packet(Binary);
        Size when Size < 28 ->
            {error, {insufficient_data, Size, "Need 28 bytes"}};
        Size when Size > 28 ->
            {error, {too_much_data, Size, "Expected 28 bytes"}}
    end;
parse(_) ->
    {error, invalid_data_type}.

% Внутренняя функция для разбора 28-байтного пакета
parse_arp_packet(<<
    HType:16,           % 2 байта - hardware type
    PType:16,           % 2 байта - protocol type
    HSize:8,            % 1 байт - hardware size
    PSize:8,            % 1 байт - protocol size
    Op:16,              % 2 байта - opcode
    SHa:6/binary,       % 6 байт - sender MAC
    SPa:4/binary,       % 4 байта - sender IP
    THa:6/binary,       % 6 байт - target MAC
    TPa:4/binary        % 4 байта - target IP
>>) ->
    % Проверка корректности размеров
    case {HSize, PSize} of
        {6, 4} ->
            % Все хорошо - возвращаем заполненный record
            #arp_packet{
                htype = HType,
                ptype = PType,
                hsize = HSize,
                psize = PSize,
                op = Op,
                sha = SHa,
                spa = SPa,
                tha = THa,
                tpa = TPa
            };
        _ ->
            {error, {invalid_address_sizes, HSize, PSize}}
    end.

% Функция для форматирования MAC-адреса
format_mac(<<A, B, C, D, E, F>>) ->
    lists:flatten(io_lib:format("~2.16.0B:~2.16.0B:~2.16.0B:~2.16.0B:~2.16.0B:~2.16.0B",
                                [A, B, C, D, E, F])).

% Функция для форматирования IP-адреса
format_ip(<<A, B, C, D>>) ->
    lists:flatten(io_lib:format("~B.~B.~B.~B", [A, B, C, D])).

% Функция для получения строки типа операции
opcode_to_string(1) -> "REQUEST (1)";
opcode_to_string(2) -> "REPLY (2)";
opcode_to_string(Op) -> lists:flatten(io_lib:format("UNKNOWN (~B)", [Op])).

% Функция вывода record
print_arp(#arp_packet{
    htype = HType,
    ptype = PType,
    hsize = HSize,
    psize = PSize,
    op = Op,
    sha = SHa,
    spa = SPa,
    tha = THa,
    tpa = TPa
}) ->
    io:format("~n=== ARP Packet ===~n"),
    io:format("Hardware type: ~B (0x~4.16.0B)~n", [HType, HType]),
    
    ProtocolStr = case PType of
        16#0800 -> " (IPv4)";
        _ -> ""
    end,
    io:format("Protocol type: 0x~4.16.0B~s~n", [PType, ProtocolStr]),
    
    io:format("Hardware size: ~B байт~n", [HSize]),
    io:format("Protocol size: ~B байт~n", [PSize]),
    io:format("Opcode: ~s~n", [opcode_to_string(Op)]),
    
    io:format("Sender MAC: ~s~n", [format_mac(SHa)]),
    io:format("Sender IP: ~s~n", [format_ip(SPa)]),
    io:format("Target MAC: ~s~n", [format_mac(THa)]),
    io:format("Target IP: ~s~n", [format_ip(TPa)]),
    
    case {HSize, PSize} of
        {6, 4} -> io:format("[OK] Формат Ethernet/IPv4~n");
        _ -> ok
    end,
    io:format("==================~n").

% Функция для тестирования нормального пакета
test_normal_packet() ->
    io:format("~n--- Тест 1: Парсинг тестового пакета ---~n"),
    case parse(?TEST_PACKET) of
        {error, Reason} ->
            io:format("[ERROR] Ошибка: ~p~n", [Reason]);
        Packet when is_record(Packet, arp_packet) ->
            io:format("[OK] Успех~n"),
            print_arp(Packet),
            
            % Проверка соответствия примеру
            io:format("~nПроверка по заданию:~n"),
            #arp_packet{spa = SPa, tpa = TPa} = Packet,
            
            SenderIP = format_ip(SPa),
            TargetIP = format_ip(TPa),
            
            io:format("Sender IP: ~s (должен быть 192.168.1.1) - ~s~n", 
                     [SenderIP, case SenderIP of "192.168.1.1" -> "[OK]"; _ -> "[ERROR]" end]),
            io:format("Target IP: ~s (должен быть 192.168.1.2) - ~s~n", 
                     [TargetIP, case TargetIP of "192.168.1.2" -> "[OK]"; _ -> "[ERROR]" end])
    end.

% Тест недостаточной длины
test_insufficient_length() ->
    io:format("~n--- Тест 2: Недостаточная длина (20 байт) ---~n"),
    SmallPacket = binary:part(?TEST_PACKET, 0, 20),
    case parse(SmallPacket) of
        {error, {insufficient_data, 20, _}} ->
            io:format("[OK] Ожидаемая ошибка о недостаточной длине~n");
        Other ->
            io:format("[ERROR] Неожиданный результат: ~p~n", [Other])
    end.

% Тест слишком большой длины
test_too_big_length() ->
    io:format("~n--- Тест 3: Слишком большая длина (32 байта) ---~n"),
    BigPacket = <<?TEST_PACKET/binary, 0,0,0,0>>,
    case parse(BigPacket) of
        {error, {too_much_data, 32, _}} ->
            io:format("[OK] Ожидаемая ошибка о слишком большой длине~n");
        Other ->
            io:format("[ERROR] Неожиданный результат: ~p~n", [Other])
    end.

% Тест неверного типа данных
test_invalid_type() ->
    io:format("~n--- Тест 4: Неверный тип данных (не бинарные) ---~n"),
    case parse(not_binary) of
        {error, invalid_data_type} ->
            io:format("[OK] Ожидаемая ошибка о неверном типе данных~n");
        Other ->
            io:format("[ERROR] Неожиданный результат: ~p~n", [Other])
    end.

% Тест ARP-ответа
test_arp_reply() ->
    io:format("~n--- Тест 5: Парсинг ARP-ответа ---~n"),
    ReplyPacket = <<
        16#00, 16#01, 16#08, 16#00, 16#06, 16#04, 16#00, 16#02,  % REPLY
        16#aa, 16#bb, 16#cc, 16#dd, 16#ee, 16#ff,                  % MAC: aa:bb:cc:dd:ee:ff
        16#c0, 16#a8, 16#01, 16#64,                                % IP: 192.168.1.100
        16#11, 16#22, 16#33, 16#44, 16#55, 16#66,                  % MAC: 11:22:33:44:55:66
        16#c0, 16#a8, 16#01, 16#01                                 % IP: 192.168.1.1
    >>,
    case parse(ReplyPacket) of
        {error, Reason} ->
            io:format("[ERROR] Ошибка: ~p~n", [Reason]);
        Packet when is_record(Packet, arp_packet) ->
            io:format("[OK] Успех~n"),
            print_arp(Packet)
    end.

% Тест неверных размеров адресов
test_invalid_sizes() ->
    io:format("~n--- Тест 6: Неверные размеры адресов ---~n"),
    BadSizePacket = <<
        16#00, 16#01, 16#08, 16#00, 16#08, 16#08, 16#00, 16#01,  % hsize=8, psize=8 (неправильно)
        16#08, 16#00, 16#27, 16#12, 16#34, 16#56, 16#C0, 16#A8,
        16#01, 16#01, 16#00, 16#00, 16#00, 16#00, 16#00, 16#00,
        16#C0, 16#A8, 16#01, 16#02
    >>,
    case parse(BadSizePacket) of
        {error, {invalid_address_sizes, 8, 8}} ->
            io:format("[OK] Ожидаемая ошибка о неверных размерах адресов~n");
        Other ->
            io:format("[ERROR] Неожиданный результат: ~p~n", [Other])
    end.

% Основная тестовая функция
test() ->
    io:format("ARP Parser Test~n"),
    io:format("================~n"),
    
    test_normal_packet(),
    test_insufficient_length(),
    test_too_big_length(),
    test_invalid_type(),
    test_arp_reply(),
    test_invalid_sizes(),
    
    io:format("~nВсе тесты завершены~n"),
    ok.